### PR TITLE
Support rocketmq protocol

### PR DESCRIPTION
**Pull request by githublaohu**
## What is the purpose of the change

XXXXX

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before
  you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address
  just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit
  in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency
  exist. If the new feature or significant change is committed, please remember to add sample
  in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure
  unit-test and integration-test pass.
- [ ] If this contribution is large, please follow
  the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0"?>
+<project
```
Comment (AlbumenJ) [inline dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml:2]: Please add ASF license header

File: dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		<dependency>
+			<groupId>org.apache.dubbo</groupId>
+			<artifactId>dubbo-registry-api</artifactId>
+			<version>${revision}</version>
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.dubbo</groupId>
+			<artifactId>dubbo-common</artifactId>
+			<version>${revision}</version>
```
Comment (AlbumenJ) [inline dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml:35]: dependencies's version  should be managed in dubbo-extensions-dependencies-bom

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml (hunk 2 above)
Comment (githublaohu) [reply dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml]: solve

File: dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml (hunk 1 above)
Comment (githublaohu) [reply dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml:17]: solve

File: dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    <parent>
+        <artifactId>dubbo-registry-extensions</artifactId>
+        <groupId>org.apache.dubbo.extensions</groupId>
+        <version>${revision}</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+	<modelVersion>4.0.0</modelVersion>
+	<artifactId>dubbo-registry-nameservice</artifactId>
+	<name>dubbo-registry-nameservice</name>
+	<url>http://maven.apache.org</url>
```
Comment (AlbumenJ) [inline dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml:30]: remove this tag

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+    private TopicList topicList;
+
+    private long timeoutMillis;
+
+    private String instanceName;
+
+    public NameServiceRegistry(URL url) {
+        super(url);
+        this.isNotRoute = !url.getParameter("route", false);
```
Comment (AlbumenJ) [inline dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java:76]: no route by default

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        this.instanceName = url.getParameter("instanceName", "nameservic-registry");
+        DefaultMQAdminExt clientConfig = new DefaultMQAdminExt();
+        clientConfig.setNamesrvAddr(url.getAddress());
+        clientConfig.setInstanceName(instanceName);
+        mqAdminExt = new DefaultMQAdminExtImpl(clientConfig, this.timeoutMillis);
+        try {
+            mqAdminExt.start();
+            this.initBeasInfo();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
Comment (AlbumenJ) [inline dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java:90]: log before throw

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        }, 1000 * 10, 3000 * 10, TimeUnit.MILLISECONDS);
+    }
+
+    private void initBeasInfo() throws Exception {
+        this.clusterInfo = this.mqAdminExt.examineBrokerClusterInfo();
+        this.topicList = this.mqAdminExt.fetchAllTopicList();
+    }
+
+    private URL createProviderURL(ServiceName serviceName, URL url, int queue) {
+        Map<String, String> parameters = url.getParameters();
```
Comment (AlbumenJ) [inline dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java:125]: replace to `url.addParameter(String, String)`

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+            return;
+        }
+        ServiceName serviceName = this.createServiceName(url);
+        try {
+            GroupList groupList = this.mqAdminExt.queryTopicConsumeByWho(serviceName.getValue());
+            if (Objects.isNull(groupList) || groupList.getGroupList().isEmpty()) {
+                return;
+            }
+        } catch (InterruptedException | MQBrokerException | RemotingException | MQClientException e) {
+            throw new RuntimeException(e);
```
Comment (AlbumenJ) [inline dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java:184]: log before throw

File: dubbo-rpc-extensions/dubbo-rpc-rocketmq/src/main/java/org/apache/dubbo/rpc/rocketmq/RocketMQInvoker.java [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+
+    @SuppressWarnings("deprecation")
+    @Override
+    protected Result doInvoke(Invocation invocation) throws Throwable {
+        RpcInvocation inv = (RpcInvocation) invocation;
+        final String methodName = RpcUtils.getMethodName(invocation);
+        inv.setAttachment(PATH_KEY, getUrl().getPath());
+        inv.setAttachment(VERSION_KEY, version);
+        // 直连
```
Comment (AlbumenJ) [inline dubbo-rpc-extensions/dubbo-rpc-rocketmq/src/main/java/org/apache/dubbo/rpc/rocketmq/RocketMQInvoker.java:103]: comment in english

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+                }
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    @Override
+    public boolean isAvailable() {
+        return false;
```
Comment (AlbumenJ) [inline dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java:158]: available if nameserver is reachable

File: dubbo-rpc-extensions/dubbo-rpc-rocketmq/src/test/java/org/apache/dubbo/rpc/rocketmq/AppTest.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
+package org.apache.dubbo.rpc.rocketmq;
```
Comment (AlbumenJ) [inline dubbo-rpc-extensions/dubbo-rpc-rocketmq/src/test/java/org/apache/dubbo/rpc/rocketmq/AppTest.java:1]: Add license header for this class

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java (hunk 6 above)
Comment (githublaohu) [reply dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java:125]: Too many parameters ， many replace。
删除太多，重写次数会很多。影响性能

File: dubbo-rpc-extensions/dubbo-rpc-rocketmq/src/test/java/org/apache/dubbo/rpc/rocketmq/AppTest.java (hunk 10 above)
Comment (githublaohu) [reply dubbo-rpc-extensions/dubbo-rpc-rocketmq/src/test/java/org/apache/dubbo/rpc/rocketmq/AppTest.java:1]: solve

File: dubbo-rpc-extensions/dubbo-rpc-rocketmq/src/main/java/org/apache/dubbo/rpc/rocketmq/RocketMQInvoker.java (hunk 8 above)
Comment (githublaohu) [reply dubbo-rpc-extensions/dubbo-rpc-rocketmq/src/main/java/org/apache/dubbo/rpc/rocketmq/RocketMQInvoker.java:103]: solve

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java (hunk 7 above)
Comment (githublaohu) [reply dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java:184]: solve

File: dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml (hunk 3 above)
Comment (githublaohu) [reply dubbo-registry-extensions/dubbo-registry-nameservice/pom.xml:30]: solve

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java (hunk 5 above)
Comment (githublaohu) [reply dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java:90]: solve

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java (hunk 6 above)
Comment (AlbumenJ) [reply dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java:125]: how about `url.addParameters(Map<String, String>)`

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+        listener.notify(urls);
+    }
+
+    void pullRoute(ServiceName serviceName, URL url, List<URL> urls) {
+        try {
+            String topic = serviceName.getValue();
+            TopicRouteData topicRouteData = this.mqAdminExt.examineTopicRouteInfo(topic);
+            Map<String, String> brokerAddrBybrokerName = new HashMap<>();
+            for (BrokerData brokerData : topicRouteData.getBrokerDatas()) {
+                brokerAddrBybrokerName.put(brokerData.getBrokerName(), brokerData.selectBrokerAddr());
```
Comment (ShannonDing) [inline dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java:206]: useless brokerAddrBybrokerName.

File: dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java (hunk 4 above)
Comment (githublaohu) [reply dubbo-registry-extensions/dubbo-registry-nameservice/src/main/java/org/apache/dubbo/registry/nameservice/NameServiceRegistry.java]: solve

**Review by AlbumenJ — APPROVED** [review]

**Pull request merged**